### PR TITLE
refactor(controller): improve Argo CD Application health check by taking conditions into account

### PR DESCRIPTION
**Pull request by hiddeco**
Fixes: #1451

Due to how Argo CD determines the "health" state of an Application, certain issues may go unnoticed at present due to it not reflecting certain types of failures as a health issue but rather as a condition in the status of an Application.

This pull request is an attempt to solve this, resulting in the following order to determine if an application (_little a_) is healthy based on the information of an Argo CD Application:

1. The Application does not have pre-flight related error conditions. If it does, it's not healthy and we should report it as "issues".
2. If we have a desired revision, and it's not synced to this revision. It's not healthy.
3. If it hasn't ran into an issue by now, the reported "health" for the Application can be trusted as the best trustworthy health we can get for the application.

### CODE REVIEW

File: internal/argocd/health.go
Comment (hiddeco) [inline internal/argocd/health.go:22]: I could use someone with Argo expertise to tell me if there are any other errors which would fall into this category.

File: internal/controller/stages/stages_test.go
Comment (krancour) [inline internal/controller/stages/stages_test.go:355]: I know you've argued in favor of this approach for some time. I maintain that there's little _technical_ difference between injecting a function directly vs injecting an implementation of an interface that _has_ only one function...... but I'll concede that this is a helluva lot easier on the eyes. 👍

Comment (hiddeco) [reply internal/controller/stages/stages_test.go:355]: To play advocate of the devil against myself, the way the other function injections could be made easier to the eyes is by creating a mocking function wrapper. E.g.

```go
func mockFunc(expectedReturn string) func(context.Context, ...) {
    return func(context.Context, ...) {
      return expectedReturn
    }
}
```

but at that point... you are almost creating the same LOC as doing mock implementations.

File: internal/argocd/health.go
Comment (krancour) [inline internal/argocd/health.go:35]: Nit: This type is a teensy bit brittle because it requires callers outside this package to have knowledge of how to initialize it. (Not that this isn't a pervasive problem in Go due to its lack of actual constructors.)

I would have a slight preference that we do this:

```
type AppHealthEvaluator interface {
	EvaluateHealth(context.Context, kargoapi.FreightReference, []kargoapi.ArgoCDAppUpdate) *kargoapi.Health
}

type appHealthEvaluator struct {
	client client.Client
}

func NewAppHealthEvaluator(c client.Client) AppHealthEvaluator {
	return &appHealthEvaluator {
		client: c,
	}
}

// etc...
```

File: internal/controller/stages/stages.go
Comment (krancour) [inline internal/controller/stages/stages.go:49]: Would go away if we did [this](https://github.com/akuity/kargo/pull/1807/files#r1558305789).

File: internal/argocd/health.go
Comment (krancour) [inline internal/argocd/health.go:101]: These are the interesting bits. Will 👀 more later this evening.

File: internal/argocd/health.go
Comment (hiddeco) [reply internal/argocd/health.go:35]: I can live with this suggestion.

The reason for the current code is because I expected the caller to be aware of the structure — which is a constructor by itself. While I did not expect there to be multiple implementations from the side of the `argocd` package — which would be reason to define an interface on this side.

This approach aligns with Go's philosophy of favoring composition over inheritance, which promotes defining interfaces on the side of the user. However, this is also frowned upon by many (including myself at times), and I do agree that your approach may be the nicer way of doing it here.

Comment (hiddeco) [reply internal/argocd/health.go]: Moved things around in 3742ef807e93b86de49b15f33c91f52c5e100a08

File: internal/argocd/health.go
Comment (hiddeco) [reply internal/argocd/health.go:165]: Maybe something like this then?

> [..] which can result in e.g. the health status result to become unreliable.

File: internal/argocd/health.go
Comment (krancour) [inline internal/argocd/health.go:30]: 🤯 that this is not in the stdlib.

**Comment by krancour** [comment]
All of this LGTM. Kudos on thinking to look at conditions _first_.

I defer to @jessesuen and/or @alexmt as better able to judge whether this seems to accomplish a comprehensive assessment of `Application` health.

I can at least verify that it solves the problems I was encountering when I opened #1451.

File: internal/argocd/health.go
Comment (jessesuen) [reply internal/argocd/health.go:22]: I cannot think of any others. We can start with this and improve it later if any others arise.

**Review by krancour — APPROVED** [review]
LGTM after conflicts are resolved.

**Pull request merged**
